### PR TITLE
1654 Fixing forecast steps in model, loss, and data loader

### LINKED ISSUE #1654: Clean-up forecast data-flow
### What happened?

In forecast mode we currently get `len(pred) = num_forecast_steps + 1`, while it should be `len(pred) = len(forecast_steps) `. 

Furthermore, `batch.get_forecast_steps()` should correspond to `num_forecast_steps`.

The loops over forecast steps in the model forward and in the loss calculator should both go from `forecast_offset` to `forecast_offset + num_forecast_steps `.

The decoder should be called after the tokens went through the forecasting engine. 

### What are the steps to reproduce the bug?

_No response_

### Hedgedoc link to logs and more information. This ticket is public, do not attach files directly.

_No response_

### PR DESCRIPTION
**Pull request by Jubeku**
## Description

This PR makes sure that: 

- [x]  In forecast mode we get `len(pred) = len(forecast_steps) + forecast_offset`.

- [x]  `batch.get_forecast_steps()` corresponds to `forecast.num_steps`.

- [x]  The loops over forecast steps in the model forward and in the loss calculator go from `forecast_offset` to `forecast_offset + num_forecast_steps` .

- [x]  The tokens go first through the forecasting engine before going through the decoder. 

## Issue Number

Closes #1654 

## Checklist before asking for review

-   [ ] I have performed a self-review of my code
-   [ ] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [ ] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [ ] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

File: src/weathergen/datasets/batch.py [hunk 1]
```diff
@@ -100,10 +100,15 @@ def get_stream_data(self, stream_name: str) -> StreamData:
...
+        for _, sdata in self.streams_data.items():
+            output_len = sdata.get_output_len()
+        return output_len
+
     def get_forecast_steps(self) -> int:
         for _, sdata in self.streams_data.items():
-            forecast_dt = sdata.get_forecast_steps()
-        return forecast_dt
+            forecast_steps = sdata.get_forecast_steps()
+        return forecast_steps
```
Comment (MatKbauer) [inline src/weathergen/datasets/batch.py]: Still don't quite understand why we have to loop over the streams here and overwrite `output_len` and `forecast_steps`, respectively. In multi-stream setup, `output_len` and `forecast_steps` will only be returned for the last stream. So why looping over streams instead of returning the last, if that is what we want?

File: src/weathergen/datasets/multi_stream_data_sampler.py [hunk 2]
```diff
@@ -476,9 +480,10 @@ def _build_stream_data(
             StreamData with source and targets masked according to view_meta
         """
 
-        # self.forecast_offset and forecast_dt are zero for pure masking
-        dt = max(1, self.forecast_offset + forecast_dt)
-        stream_data = StreamData(base_idx, num_steps_input, dt, self.num_healpix_cells)
+        dt = self._get_output_length(forecast_dt)
+        stream_data = StreamData(
+            base_idx, num_steps_input, dt, forecast_dt, self.forecast_offset, self.num_healpix_cells
+        )
```
Comment (MatKbauer) [inline src/weathergen/datasets/multi_stream_data_sampler.py:493]: Can we rename
- `num_steps_input` to `num_input_steps`,
- `dt` to `num_forecast_steps`, and
- `forecast_dt` to `forecast_steps`?

File: src/weathergen/datasets/stream_data.py [hunk 3]
```diff
@@ -371,7 +389,13 @@ def get_forecast_steps(self) -> int:
         """
         Get number of forecast steps
         """
-        return self.forecast_steps
+        return self.forecast_idxs
```
Comment (MatKbauer) [inline src/weathergen/datasets/stream_data.py]: Rename function from `get_forecast_steps()` to `get_forecast_idxs` and adapt function header (side note, I'd prefer to use `idcs` over `idxs`, but we have `idxs` at various places in the code)

File: src/weathergen/datasets/batch.py (hunk 1 above)
Comment (MatKbauer) [reply src/weathergen/datasets/batch.py]: In line with comments below, I'd advocate for `get_num_forecast_steps()` instead of `get_forecast_steps()`

File: src/weathergen/datasets/stream_data.py [hunk 4]
```diff
@@ -70,28 +87,29 @@ def __init__(self, idx: int, input_steps: int, forecast_steps: int, healpix_cell
         self.mask_value = 0.0
 
         self.input_steps = input_steps
-        self.forecast_steps = forecast_steps
+        self.output_steps = output_steps
```
Comment (MatKbauer) [inline src/weathergen/datasets/stream_data.py:84]: Can we use `self.num_output_steps` and `self.num_input_steps` to be consistent with `num_forecast_steps` in the `multi_stream_data_sampler`?

File: src/weathergen/datasets/stream_data.py [hunk 5]
```diff
@@ -371,7 +389,13 @@ def get_forecast_steps(self) -> int:
...
         Get number of forecast steps
         """
-        return self.forecast_steps
+        return self.forecast_idxs
+
+    def get_output_len(self) -> int:
+        """
+        Get length of output
+        """
+        return self.output_steps
```
Comment (MatKbauer) [inline src/weathergen/datasets/stream_data.py]: `return self.num_output_steps`

File: src/weathergen/model/model.py [hunk 6]
```diff
@@ -567,43 +575,30 @@ def forward(
...
-                register_tokens=tokens[:, : self.register_token_idx],
-                class_token=tokens[:, self.register_token_idx : self.class_token_idx],
-                patch_tokens=tokens[:, self.class_token_idx :],
-                z_pre_norm=None,
-            )
-            output.add_latent_prediction(fstep, "latent_state", latent_state)
-
-        # prediction for final step
-        output = self.predict(model_params, batch.get_forecast_steps(), tokens, batch, output)
+        if len(batch.get_forecast_steps()) > 0:
```
Comment (MatKbauer) [inline src/weathergen/model/model.py]: `batch.get_num_forecast_steps()`

File: src/weathergen/train/loss_modules/loss_module_physical.py [hunk 7]
```diff
@@ -219,23 +217,28 @@ def compute_loss(self, preds: dict, targets: dict, metadata) -> LossValues:
 
             stream_loss_weight, weights_channels = self._get_weights(stream_info)
 
-            fstep_loss_weights = self._get_fstep_weights(targets.num_forecast_steps)
+            # TODO: make nicer
+            fstep_loss_weights = self._get_fstep_weights(len(targets.forecast_steps))
```
Comment (MatKbauer) [inline src/weathergen/train/loss_modules/loss_module_physical.py]: Can we rename `targets.forecast_steps` to `targets.forecast_idxs` (or `forecast_idcs` :p)?

File: src/weathergen/train/loss_modules/loss_module_physical.py [hunk 8]
```diff
@@ -219,23 +217,28 @@ def compute_loss(self, preds: dict, targets: dict, metadata) -> LossValues:
...
+            if len(targets.physical) - len(targets.forecast_steps) > 0:
+                fstep_loss_weights.insert(0, None)
 
             loss_fsteps = torch.tensor(0.0, device=self.device, requires_grad=True)
             ctr_fsteps = 0
 
-            for fstep in range(self.forecast_offset, targets.num_forecast_steps):
+            # TODO loop directly through preds and targets
+            fstep_idxs = [0] if len(targets.forecast_steps) == 0 else targets.forecast_steps
```
Comment (MatKbauer) [inline src/weathergen/train/loss_modules/loss_module_physical.py]: `targets.forecast_idxs`

File: src/weathergen/train/target_and_aux_module_base.py [hunk 9]
```diff
@@ -13,19 +13,49 @@
...
+
 
 @dataclasses.dataclass
 class TargetAuxOutput:
     """
     A dataclass to encapsulate the TargetAndAuxCalculator output and give a clear API.
     """
 
-    num_forecast_steps: int
+    forecast_steps: list
```
Comment (MatKbauer) [inline src/weathergen/train/target_and_aux_module_base.py]: I suppose these are `forecast_idxs`

File: src/weathergen/datasets/batch.py (hunk 1 above)
Comment (Jubeku) [inline src/weathergen/datasets/batch.py]: I am still a bit concerned about the for loops here in which the variable is overwritten at each iteration. I assume both `forecast_steps` and `output_len` are constant within a batch across all streams?! If yes we can probably use it to handle it better.

File: src/weathergen/train/target_and_aux_module_base.py [hunk 10]
```diff
@@ -13,19 +13,49 @@
...
+    latent: list[dict[str, torch.Tensor | LatentState]]
     aux_outputs: dict[str, torch.Tensor]
 
+    def __init__(self, len_target: int, forecast_steps: list) -> None:
+        self.forecast_steps = forecast_steps
+        self.physical = [{} for _ in range(len_target)]
+        self.latent = [{} for _ in range(len_target)]
+        self.aux_outputs = {}
+
+    def add_physical_target(self, fstep: int, stream_name: StreamName, pred: torch.Tensor) -> None:
```
Comment (MatKbauer) [inline src/weathergen/train/target_and_aux_module_base.py]: `fidx`

File: src/weathergen/train/target_and_aux_module_base.py [hunk 11]
```diff
@@ -13,19 +13,49 @@
...
-    num_forecast_steps: int
+    forecast_steps: list
 
-    physical: dict[str, torch.Tensor]
-    latent: dict[str, torch.Tensor]
+    physical: list[dict[StreamName, torch.Tensor]]
+    latent: list[dict[str, torch.Tensor | LatentState]]
     aux_outputs: dict[str, torch.Tensor]
 
+    def __init__(self, len_target: int, forecast_steps: list) -> None:
```
Comment (MatKbauer) [inline src/weathergen/train/target_and_aux_module_base.py]: `forecast_idxs`

File: src/weathergen/train/target_and_aux_module_base.py [hunk 12]
```diff
@@ -60,18 +90,19 @@ def update_state_pre_backward(self, istep, batch, model, **kwargs):
...
         # TODO: properly retrieve/define these
         stream_names = [k for k, _ in batch.samples[0].streams_data.items()]
         forecast_steps = batch.get_forecast_steps()
 
+        targets = TargetAuxOutput(batch.get_output_len(), forecast_steps)
+
         # collect all targets, concatenating across batch dimension since this is also how it
         # happens for predictions in the model
-        targets = {}
+        fstep_idxs = [0] if len(forecast_steps) == 0 else forecast_steps
```
Comment (MatKbauer) [inline src/weathergen/train/target_and_aux_module_base.py]: How about using `fidxs` here and `fidx` in the forecast loop below?

File: src/weathergen/train/target_and_aux_module_base.py [hunk 13]
```diff
@@ -60,18 +90,19 @@ def update_state_pre_backward(self, istep, batch, model, **kwargs):
     def update_state_post_opt_step(self, istep, batch, model, **kwargs):
         return
 
-    def compute(self, istep, batch, *args, **kwargs) -> TargetAuxOutput:
+    def compute(self, bidx, batch, model_params, model) -> TargetAuxOutput:
         # TODO: properly retrieve/define these
         stream_names = [k for k, _ in batch.samples[0].streams_data.items()]
         forecast_steps = batch.get_forecast_steps()
```
Comment (MatKbauer) [inline src/weathergen/train/target_and_aux_module_base.py]: `forecast_idxs = batch.get_forecast_idxs()`

File: src/weathergen/utils/validation_io.py [hunk 14]
```diff
@@ -40,11 +40,12 @@ def write_output(
...
 
-    window_offset_prediction = val_cfg.get("window_offset_prediction", 0)
+    forecast_offset = val_cfg.get("forecast", {}).get("offset", 0)
     forecast_steps = max(1, val_cfg.get("forecast", {}).get("num_steps", 1))
     targets_lens = []
-    # for fstep in range(window_offset_prediction, forecast_steps + 1):
-    for fstep in range(window_offset_prediction, forecast_steps):
+    # TODO why does this stop at forecast_steps? Maybe explains #1657
+    # for fstep in range(forecast_offset, forecast_steps + 1):
+    for fstep in range(forecast_offset, forecast_steps):
```
Comment (MatKbauer) [inline src/weathergen/utils/validation_io.py]: Crucial to keep in mind. Maybe have to extend this to `+1` here or iterate over `forecast_idxs`

**Review by MatKbauer — COMMENTED** [review]
Looks good so far. I'd like to make one major suggestion: Can we rename `forecast_steps` to `forecast_idxs` (or `forecast_idcs`) to be very explicit in our naming? I would even go down to changing `fstep` to `fidx` then.

Most of my comments in the review relate to this suggested change in naming convention.

File: src/weathergen/train/target_and_aux_module_base.py (hunk 12 above)
Comment (Jubeku) [reply src/weathergen/train/target_and_aux_module_base.py]: I also considered something like target_idxs to make clear this not only holds for forecasting. But `target_idx` is also used later in the loss module to refer to the source-target correspondence so it invites risky confusions.

File: src/weathergen/datasets/multi_stream_data_sampler.py [hunk 15]
```diff
@@ -281,12 +281,14 @@ def reset(self):
         self.rng = np.random.default_rng(self.data_loader_rng_seed)
 
         fsm = (
-            self.forecast_steps[min(self.mini_epoch, len(self.forecast_steps) - 1)]
```
Comment (Jubeku) [inline src/weathergen/datasets/multi_stream_data_sampler.py:284]: @clessig, why is `len(self.forecast_steps)` used here?

File: src/weathergen/datasets/multi_stream_data_sampler.py [hunk 16]
```diff
@@ -317,13 +319,16 @@ def reset(self):
...
         if self.forecast_policy is None:
-            self.perms_forecast_dt = np.zeros(len_dt_samples, dtype=np.int64)
+            self.perms_num_forecast_steps = np.zeros(len_dt_samples, dtype=np.int64)
         elif self.forecast_policy == "fixed" or self.forecast_policy == "sequential":
-            self.perms_forecast_dt = fsm * np.ones(len_dt_samples, dtype=np.int64)
+            self.perms_num_forecast_steps = fsm * np.ones(len_dt_samples, dtype=np.int64)
         elif self.forecast_policy == "random" or self.forecast_policy == "sequential_random":
             # randint high=one-past
-            self.perms_forecast_dt = self.rng.integers(
-                low=self.forecast_steps.min(), high=fsm + 1, size=len_dt_samples, dtype=np.int64
```
Comment (Jubeku) [inline src/weathergen/datasets/multi_stream_data_sampler.py:326]: Why is here `high=fsm+1`?

**Review by MatKbauer — APPROVED** [review]
Works well for me, when testing `forecasting` with `offset: 0` or `1` and different `num_steps`. I launched a 2-node forecast pre-training (2 forecast steps) and would like to quickly verify 8 forecast steps fine-tuning subsequently.

**Comment by clessig** [comment]
Closed via #1690
